### PR TITLE
Primitive collection events

### PR DESCRIPTION
**Pull request by chris-cooper**
This adds `primitiveAdded` and `primitiveRemoved` events to `PrimitiveCollection` as discussed in #9815 .  The implementation is similar to [`DataSourceCollection`](https://github.com/CesiumGS/cesium/blob/main/packages/engine/Source/DataSources/DataSourceCollection.js#L33-L57).

This change has been in our fork for some time.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @chris-cooper!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/primitive-collection-events/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.
* :grey_question: Unit tests were not updated.
    * Make sure you've [updated tests](https://github.com/CesiumGS/cesium/tree/main/Documentation/Contributors/TestingGuide) to reflect your changes, added tests for any new code, and ran the code coverage tool.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.

File: packages/engine/Specs/Scene/PrimitiveCollectionSpec.js [hunk 1]
```diff
@@ -649,6 +649,42 @@ describe(
...
     });
+
+    it("fires an event on primitive added", function () {
+      const primitiveAdded = primitives.primitiveAdded;
+
+      const onPrimitiveAddedSpy = jasmine.createSpy("event");
+      primitiveAdded.addEventListener(onPrimitiveAddedSpy);
+      expect(onPrimitiveAddedSpy).not.toHaveBeenCalled();
+      primitives.add(createLabels());
+      expect(onPrimitiveAddedSpy).toHaveBeenCalled();
```
Comment (ggetz) [inline packages/engine/Specs/Scene/PrimitiveCollectionSpec.js:660]: I would suggest using `toHaveBeenCalledWith` here and below to ensure the primitives were indeed passed to the callback.

**Review by ggetz — COMMENTED** [review]
Thanks @chris-cooper! I think this is a sensible and low-impact addition to the Primitive API.

I just have one suggestion to make the unit tests a bit more thorough. Otherwise, the code is looking good!

File: packages/engine/Specs/Scene/PrimitiveCollectionSpec.js [hunk 2]
```diff
@@ -649,6 +649,51 @@ describe(
...
+      primitiveRemoved.addEventListener(onPrimitiveRemovedSpy);
+      expect(onPrimitiveRemovedSpy).not.toHaveBeenCalled();
+      primitives.add(p);
+      expect(onPrimitiveRemovedSpy).not.toHaveBeenCalled();
+      primitives.remove(p);
+      expect(onPrimitiveRemovedSpy).toHaveBeenCalledOnceWith(p);
+    });
+
+    it("fires events on primitive remove all", function () {
+      const collection = new PrimitiveCollection({ destroyPrimitives: false });
```
Comment (chris-cooper) [inline packages/engine/Specs/Scene/PrimitiveCollectionSpec.js:678]: I'm manually destroying as the `toHaveBeenCalledWith()` test was triggering a 'this primitive has been destroyed' error.

**Comment by chris-cooper** [comment]
Good idea thanks @ggetz .  It should be ready for another look now.

File: packages/engine/Specs/Scene/PrimitiveCollectionSpec.js (hunk 2 above)
Comment (ggetz) [reply packages/engine/Specs/Scene/PrimitiveCollectionSpec.js:678]: Should this be noted in the `primitiveRemoved` docs? This seems something likely for devs to run into during development as well.

**Comment by ggetz** [comment]
Thanks for the update @chris-cooper! I have one more comment based on your explanation above.

**Comment by chris-cooper** [comment]
@ggetz I've added a note to the `primitiveRemoved` docs.  I've also conformed the ordering of operations between `remove()` and `removeAll()`.  It can lead to some hard to track down issues if `destroy()` is called immediately after the event handler as I found when debugging the spec.  Now in both cases `destroy()` is called before.  If the user wants to do something with the primitive when it's removed prior to being destroyed then they can handle the destroy themselves.

**Comment by ggetz** [comment]
Great, thanks @chris-cooper!

**Pull request merged**